reactos/drivers/storage/storahci/storahci.c

```diff
@@ -0,0 +1,745 @@
+                                (*(ConfigInfo->AccessRanges))[index].RangeLength,
+                                (BOOLEAN)!(*(ConfigInfo->AccessRanges))[index].RangeInMemory);
+                break;
+            }
+        }
+    }
+
+    if (abar == NULL){
+        StorPortDebugPrint(0, "\tabar == NULL\n");
+        return SP_RETURN_ERROR; // corrupted information supplied
+    }
+
+    adapterExtension->ABAR_Address = abar;
+    adapterExtension->CAP = StorPortReadRegisterUlong(adapterExtension, &abar->CAP);
+    adapterExtension->CAP2 = StorPortReadRegisterUlong(adapterExtension, &abar->CAP2);
+    adapterExtension->Version = StorPortReadRegisterUlong(adapterExtension, &abar->VS);
+    adapterExtension->LastInterruptPort = -1;
+
+    // 10.1.2
+    // 1. Indicate that system software is AHCI aware by setting GHC.AE to ‘1’.
+    // 3.1.2 -- AE bit is read-write only if CAP.SAM is '0'
+    ghc = StorPortReadRegisterUlong(adapterExtension, &abar->GHC);
+    // AE := Highest Significant bit of GHC
+    if ((ghc & AHCI_Global_HBA_CONTROL_AE) != 0)//Hmm, controller was already in power state
+    {
+        // reset controller to have it in known state
+        StorPortDebugPrint(0, "\tAE Already set, Reset()\n");
+        if (!AhciAdapterReset(adapterExtension)){
+            StorPortDebugPrint(0, "\tReset Failed!\n");
+            return SP_RETURN_ERROR;// reset failed
+        }
+    }
+
+    ghc = AHCI_Global_HBA_CONTROL_AE;// only AE=1
+    StorPortWriteRegisterUlong(adapterExtension, &abar->GHC, ghc);
+
+    adapterExtension->IS = &abar->IS;
+    adapterExtension->PortImplemented = StorPortReadRegisterUlong(adapterExtension, &abar->PI);
+
+    if (adapterExtension->PortImplemented == 0){
+        StorPortDebugPrint(0, "\tadapterExtension->PortImplemented == 0\n");
+        return SP_RETURN_ERROR;
+    }
+
+    ConfigInfo->MaximumTransferLength = 128 * 1024;//128 KB
+    ConfigInfo->NumberOfPhysicalBreaks = 0x21;
+    ConfigInfo->MaximumNumberOfTargets = 1;
+    ConfigInfo->MaximumNumberOfLogicalUnits = 1;
+    ConfigInfo->ResetTargetSupported = TRUE;
+    ConfigInfo->NumberOfBuses = MAXIMUM_AHCI_PORT_COUNT;
+    ConfigInfo->SynchronizationModel = StorSynchronizeFullDuplex;
+    ConfigInfo->ScatterGather = TRUE;
+
+    // Turn IE -- Interrupt Enabled
+    ghc |= AHCI_Global_HBA_CONTROL_IE;
+    StorPortWriteRegisterUlong(adapterExtension, &abar->GHC, ghc);
+
+    // allocate necessary resource for each port
+    if (!AhciAllocateResourceForAdapter(adapterExtension, ConfigInfo)){
+        StorPortDebugPrint(0, "\tAhciAllocateResourceForAdapter() == FALSE\n");
+        return SP_RETURN_ERROR;
+    }
+
+    for (index = 0; index < MAXIMUM_AHCI_PORT_COUNT; index++)
+    {
+        if ((adapterExtension->PortImplemented & (0x1<<index)) != 0)
+            AhciPortInitialize(&adapterExtension->PortExtension[index]);
+    }
+
+    return SP_RETURN_FOUND;
+}// -- AhciHwFindAdapter();
+

```

Comment:
I stopped looking for code style issues at this point, because I would be repeating the same things :)